### PR TITLE
Refactoring of Cypher compatibility

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

**Comment by henriknyman** [comment]
[Refactoring of Cypher compatibility](https://trello.com/c/uH1A6RPR/232-refactoring-of-cypher-compatibility)

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlannerCache.scala [hunk 1]
```diff
@@ -43,34 +43,34 @@ class PlannerFactory(graph: GraphDatabaseQueryService, kernelAPI: KernelAPI, ker
...
-
 }
 
 class PlannerCache(factory: PlannerFactory)  {
-  private val cache_v2_3 = new mutable.HashMap[PlannerSpec_v2_3, CompatibilityFor2_3]
-  private val cache_v3_1 = new mutable.HashMap[PlannerSpec_v3_1, CompatibilityFor3_1]
-  private val cache_v3_2 = new mutable.HashMap[PlannerSpec_v3_2, CompatibilityFor3_2]
+  private val cache_v2_3 = new mutable.HashMap[PlannerSpec_v2_3, v2_3.Compatibility]
+  private val cache_v3_1 = new mutable.HashMap[PlannerSpec_v3_1, v3_1.Compatibility]
+  private val cache_v3_2 = new mutable.HashMap[PlannerSpec_v3_2, v3_2.Compatibility]
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlannerCache.scala:70]: Just curious; what is this cache? Should we use Caffeine here as well?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/RewindableExecutionResult.scala [hunk 2]
```diff
@@ -76,7 +78,7 @@ object RewindableExecutionResult {
   private def compatibility(inner: v3_1.executionplan.InternalExecutionResult, planner: v3_1.PlannerName, runtime: v3_1.RuntimeName): InternalExecutionResult = {
     val result: v3_1.executionplan.InternalExecutionResult = inner match {
       case other: v3_1.PipeExecutionResult =>
-        exceptionHandlerFor2_3.runSafely {
+        exceptionHandlerFor3_1.runSafely {
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/RewindableExecutionResult.scala:81]: That's a bugfix!

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/NewPlannerTestSupport.scala [hunk 3]
```diff
@@ -285,9 +286,9 @@ trait NewPlannerTestSupport extends CypherTestSupport {
   private def rewindableResult(result: ExecutionResult): InternalExecutionResult = {
     result match {
       case e: ClosingExecutionResult => e.inner match {
-        case _: ExecutionResultWrapperFor3_2 => RewindableExecutionResult(e)
-        case _: ExecutionResultWrapperFor3_1 => RewindableExecutionResult(e)
-        case _: ExecutionResultWrapperFor2_3 => RewindableExecutionResult(e)
+        case _: ExecutionResultWrapper => RewindableExecutionResult(e)
+        case _: v3_1.ExecutionResultWrapper => RewindableExecutionResult(e)
+        case _: v2_3.ExecutionResultWrapper => RewindableExecutionResult(e)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/NewPlannerTestSupport.scala:291]: This reads much better I think. Good improvement!

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_2/exceptionHandler.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  def invalidArgumentException(message: String, cause: Throwable) = new InvalidArgumentException(message, cause)
+
+  def mergeConstraintConflictException(message: String, cause: Throwable) = new MergeConstraintConflictException(message, cause)
+
+  def internalException(message: String, cause: Exception) = new InternalException(message, cause)
+
+  def loadCsvStatusWrapCypherException(extraInfo: String, cause: InternalCypherException) =
+    new LoadCsvStatusWrapCypherException(extraInfo, cause.mapToPublic(exceptionHandler))
+
+  def loadExternalResourceException(message: String, cause: Throwable) = throw new LoadExternalResourceException(message, cause)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_2/exceptionHandler.scala:50]: Not a critique of this PR, but how come some of these functions are returning objects, and some are throwing them?

**Review by Mats-SX — APPROVED** [review]
Looks very good!

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlannerCache.scala (hunk 1 above)
Comment (systay) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlannerCache.scala:70]: By just looking at the code - these are very small maps that are not emptied. I think it makes sense to move from our LRU caches to Caffeine, but this is not the same thing.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/RewindableExecutionResult.scala (hunk 2 above)
Comment (systay) [reply community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/RewindableExecutionResult.scala:81]: It is. It's test-only and I don't think we ever use the exception handling in the tests, so I don't think it's important to backport though.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_2/exceptionHandler.scala (hunk 4 above)
Comment (systay) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_2/exceptionHandler.scala:50]: That is probably a bug. I'll look at older branches and see if the same thing is happening. Good catch!

**Pull request merged**